### PR TITLE
[servicescm] Service dependency cycle detection

### PR DESCRIPTION
**Pull request by saifshaikh48**
This PR adds cycle detection to the validation that occurs
on the services defined in `windows-services` ConfigMaps.
A dependency chain with cycles is considered invalid.

### CODE REVIEW

**Comment by selansen** [comment]
Thanks for the review discussion. 

/approve

File: pkg/servicescm/servicescm.go [hunk 1]
```diff
@@ -278,6 +283,48 @@ func (s *Service) hasDependency(possibleDependencies []Service) bool {
...
+	}
+
+	visited := make(map[string]struct{})
+	currentPath := make(map[string]struct{})
+	for _, svc := range services {
+		// Check if helper has already been called on this service to prevent duplicate calls
+		if _, present := visited[svc.Name]; !present {
+			if cycleDetected := svc.hasCycle(servicesMap, visited, currentPath); cycleDetected {
+				return true
+			}
```
Comment (jrvaldes) [inline pkg/servicescm/servicescm.go:302]:
```suggestion
			if svc.hasCycle(servicesMap, visited, currentPath) {
				return true
			}
```

File: pkg/servicescm/servicescm.go [hunk 2]
```diff
@@ -278,6 +283,48 @@ func (s *Service) hasDependency(possibleDependencies []Service) bool {
 	return false
 }
 
+// hasCycle detects cycles between any of the given services by traversing the resulting dependency graph.
+// Wrapper for Service.hasCycle to handle disconnected graphs
```
Comment (jrvaldes) [inline pkg/servicescm/servicescm.go:287]: Can you add more color around the "disconnected graphs" case?

File: pkg/servicescm/servicescm_test.go [hunk 3]
```diff
@@ -153,7 +153,7 @@ func TestValidateDependencies(t *testing.T) {
 				{
 					Name:         "test-controller-service-2",
 					Command:      "C:\\test-controller-service-2",
-					Dependencies: []string{"test-controller-service-2"},
+					Dependencies: []string{"test-controller-service"},
```
Comment (jrvaldes) [inline pkg/servicescm/servicescm_test.go:156]: Is this change related to the cyclical dependency test?

File: pkg/servicescm/servicescm_test.go [hunk 4]
```diff
@@ -238,6 +238,60 @@ func TestValidateDependencies(t *testing.T) {
...
 		},
+		{
+			name: "Cyclical dependency structure",
+			input: []Service{
+				{
+					Name:         "new-bootstrap-service",
+					Command:      "C:\\new-service",
+					Dependencies: []string{"test-controller-service"},
+					Bootstrap:    false,
+					Priority:     0,
```
Comment (jrvaldes) [inline pkg/servicescm/servicescm_test.go:249]: It may be easier to read if you use the priority value to fill all the fields. For example:

```
					Name:         "service-0",
					Command:      "C:\\service-0",
					Dependencies: []string{"service-1"},
					Bootstrap:    false,
					Priority:     0,
```

File: pkg/servicescm/servicescm.go [hunk 5]
```diff
@@ -278,6 +283,48 @@ func (s *Service) hasDependency(possibleDependencies []Service) bool {
...
+		if _, present := visited[svc.Name]; !present {
+			if cycleDetected := svc.hasCycle(servicesMap, visited, currentPath); cycleDetected {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// hasCycle uses depth-first traversal to check for cycles in the service dependency graph, using s as the source node
```
Comment (jrvaldes) [inline pkg/servicescm/servicescm.go:308]: Did you considered a [tortoise and hare](https://en.wikipedia.org/wiki/Cycle_detection#Floyd's_tortoise_and_hare)-like approach for finding the cycle?

File: pkg/servicescm/servicescm.go [hunk 6]
```diff
@@ -278,6 +283,48 @@ func (s *Service) hasDependency(possibleDependencies []Service) bool {
...
+	visited[s.Name] = struct{}{}
+	currentPath[s.Name] = struct{}{}
+
+	for _, dependencyName := range s.Dependencies {
+		if _, present := currentPath[dependencyName]; present {
+			// Cycle detected if service already exists in current path
+			return true
+		}
+		if _, present := visited[dependencyName]; !present {
+			return servicesMap[dependencyName].hasCycle(servicesMap, visited, currentPath)
```
Comment (jrvaldes) [inline pkg/servicescm/servicescm.go:320]: In the best-case scenario (no cycles) `servicesMap` and `visited` have the same keys, consider using `servicesMap` as the primary lookup entity so that you reduce the extra space complexity by removing the `visited` set.

File: pkg/servicescm/servicescm_test.go [hunk 7]
```diff
@@ -238,6 +238,60 @@ func TestValidateDependencies(t *testing.T) {
...
+			expectedErr: true,
+		},
+		{
+			name: "Disconnect cyclical dependency structure",
+			input: []Service{
+				{
+					Name:         "new-bootstrap-service",
+					Command:      "C:\\new-service",
+					Dependencies: []string{},
+					Bootstrap:    false,
```
Comment (jrvaldes) [inline pkg/servicescm/servicescm_test.go:275]: Based on the name (`new-bootstrap-service`), shouldn't this be `true`?

File: pkg/servicescm/servicescm_test.go [hunk 8]
```diff
@@ -238,6 +238,60 @@ func TestValidateDependencies(t *testing.T) {
...
+					Command:      "C:\\test-controller-service-2",
+					Dependencies: []string{"new-bootstrap-service"},
+					Bootstrap:    false,
+					Priority:     2,
+				},
+			},
+			expectedErr: true,
+		},
+		{
+			name: "Disconnect cyclical dependency structure",
```
Comment (jrvaldes) [inline pkg/servicescm/servicescm_test.go:269]: Why disconnected? IMO, this test is equivalent to the first one (`Cyclical dependency structure`).

File: pkg/servicescm/servicescm_test.go [hunk 9]
```diff
@@ -238,6 +238,60 @@ func TestValidateDependencies(t *testing.T) {
...
+				{
+					Name:         "test-controller-service-2",
+					Command:      "C:\\test-controller-service-2",
+					Dependencies: []string{"test-controller-service"},
+					Bootstrap:    false,
+					Priority:     2,
+				},
+			},
+			expectedErr: true,
+		},
```
Comment (jrvaldes) [inline pkg/servicescm/servicescm_test.go:294]: Can a service depend on itself? Is this case covered?

```
				{
					Name:         "test-service",
					Command:      "C:\\test-service",
					Dependencies: []string{"test-service"},
					Bootstrap:    false,
					Priority:     0,
				},
```

**Review by jrvaldes — COMMENTED** [review]
Nice work @saifshaikh48, added a few comments. PTAL

File: pkg/servicescm/servicescm.go [hunk 10]
```diff
@@ -278,6 +283,48 @@ func (s *Service) hasDependency(possibleDependencies []Service) bool {
 	return false
 }
 
+// hasCycle detects cycles between any of the given services by traversing the resulting dependency graph.
+// Wrapper for Service.hasCycle to handle disconnected graphs
+func hasCycle(services []Service) bool {
```
Comment (sebsoto) [inline pkg/servicescm/servicescm.go:288]: please rename this to something equivalent, would prefer to not have multiple functions with the same name if they are not due to an interface/generics

`validateCycles` perhaps? That way you can also return an error and state which service in particular has the cycle.

File: pkg/servicescm/servicescm.go (hunk 5 above)
Comment (sebsoto) [reply pkg/servicescm/servicescm.go:308]: +1 on a solution that does not use recursion.
Recursion can feel elegant, but we should prefer simplicity when we can, especially since we don't need these functions to be performant.
In many cases recursion makes things harder to read and visualize, especially when anticipating edge cases. This gets increasingly worse as the requirements of our code evolves. Granted its early, but its taking me a long time to figure out whats going on here, compared to if this was done with some simple loops.

File: pkg/servicescm/servicescm_test.go (hunk 9 above)
Comment (saifshaikh48) [reply pkg/servicescm/servicescm_test.go:294]: A service that depends on itself is invalid as it is considered a cycle. I'll add a unit test around this

File: pkg/servicescm/servicescm_test.go (hunk 7 above)
Comment (saifshaikh48) [reply pkg/servicescm/servicescm_test.go:275]: I'll fix this, though it has no effect on the test case

File: pkg/servicescm/servicescm_test.go (hunk 3 above)
Comment (saifshaikh48) [reply pkg/servicescm/servicescm_test.go:156]: Yes, this test case needed to be modified since a service cannot depend on itself.

File: pkg/servicescm/servicescm.go (hunk 2 above)
Comment (saifshaikh48) [reply pkg/servicescm/servicescm.go:287]: Consider 3 services: A, B, and C
- A has no dependencies, B depends on C, C depends on B. 
- The resulting graph has 2 distinct, disconnected components: one is {A} and the other is {B,C}

As such, we cannot chose any node to call Service.hasCycles() on. If we simply did A.hasCycles(), it would return false despite the B<-->C cycle. So we need to call the method once on each subgraph.

File: pkg/servicescm/servicescm.go (hunk 5 above)
Comment (saifshaikh48) [reply pkg/servicescm/servicescm.go:308]: I'm going to push back on this after exploring the iterative solution. The iterative solution is quite hard to follow itself and requires a stack (not hard to implement, just adds avoidable custom logic)

Comment (sebsoto) [reply pkg/servicescm/servicescm.go:305]: okay, thats fine with me, I'll remove the hold
thanks for exploring the other solution

**Comment by selansen** [comment]
I did review the code with @saifshaikh48  to go through DFP graph loops. It gave me the opportunity to refresh my memory.
But I am +1 for making it simpler. 
@saifshaikh48 do you still want to merge this PR and rework a more straightforward solution later?
As this is not a blocker for WICD work, I suggest you keep this PR open and work on the alternate solution rather than pushing and then replacing it with an alternate solution.

File: pkg/servicescm/servicescm.go (hunk 6 above)
Comment (saifshaikh48) [reply pkg/servicescm/servicescm.go:317]: After revisiting this again, I was able to find a way to consolidate the 2 maps using a variant of the [color-coded backtracking](https://www.geeksforgeeks.org/detect-cycle-direct-graph-using-colors/) algorithm. Thanks for the suggestion Jose.

**Comment by saifshaikh48** [comment]
@selansen I did a fair bit of investigation here and tried to write up a few alternate solutions. A recursive DFS strategy is definitely the best suited/simplest for this problem.

**Comment by jrvaldes** [comment]
LGTM

**Comment by mansikulkarni96** [comment]
LGTM will wait for @sebsoto to give final lgtm
